includes/UserMailer.php

```diff
@@ -434,114 +412,189 @@ protected static function quotedPrintableCallback( $matches ) {
-			$sendEmail = false;
-			// Only send notification for non minor edits, unless $wgEnotifMinorEdits
-			if ( !$minorEdit || ( $wgEnotifMinorEdits && !$editor->isAllowed( 'nominornewtalk' ) ) ) {
-				$isUserTalkPage = ( $title->getNamespace() == NS_USER_TALK );
-				if ( $wgEnotifUserTalk && $isUserTalkPage && $this->canSendUserTalkEmail( $editor, $title, $minorEdit ) ) {
-					$sendEmail = true;
-				}
-			}
+	/**
+	 * Add a timeout to the watchlist email block
+	 */
+	private function getTimeOutSql() {
+		global $wgEnableWatchlistNotificationTimeout, $wgWatchlistNotificationTimeout;
+
+		if ( !empty( $this->otherParam['notisnull'] ) ) {
+			$notificationTimeoutSql = "1";
+		} elseif ( !empty( $wgEnableWatchlistNotificationTimeout ) && isset( $WatchlistNotificationTimeout ) ) {
+			$blockTimeout = wfTimestamp( TS_MW, wfTimestamp( TS_UNIX, $this->timestamp ) - intval( $wgWatchlistNotificationTimeout ) );
+			$notificationTimeoutSql = "wl_notificationtimestamp IS NULL OR wl_notificationtimestamp < '$blockTimeout'";
+		} else {
+			$notificationTimeoutSql = 'wl_notificationtimestamp IS NULL';
 		}
 
-		if ( !$sendEmail ) {
-			return;
+		return $notificationTimeoutSql;
+	}
+
+	private function getWatchersToNotify( $notificationTimeoutSql ) {
+		$watchers = [];
+		$dbw = wfGetDB( DB_MASTER );
+		$res = $dbw->select( [ 'watchlist' ],
+			[ 'wl_user' ],
+			[
+				'wl_title' => $this->title->getDBkey(),
+				'wl_namespace' => $this->title->getNamespace(),
+				'wl_user != ' . intval( $this->editor->getID() ),
+				$notificationTimeoutSql
+			], __METHOD__
+		);
+
+		foreach ( $res as $row ) {
+			$watchers[] = $row->wl_user;
+		}
+
+		return $watchers;
+	}
+
+	/**
+	 * Update wl_notificationtimestamp for all watching users except the editor
+	 */
+	private function updateWatchedItem( array $watchers ) {
+		$wl = WatchedItem::fromUserTitle( $this->editor, $this->title );
+		$wl->updateWatch( $watchers, $this->timestamp );
+	}
+
+	/**
+	 * Check if there are either 1.) users watching the page, or 2.) the page which was
+	 * edited is a User Talk page and the owner of that page wants to receive notifications
+	 * when their User Talk page is changed.
+	 * @param array $watchers
+	 * @return bool
+	 */
+	private function shouldSendEmail( array $watchers ) {
+		if ( $this->thereAreUsersWatchingPage( $watchers ) ) {
+			return true;
+		}
+
+		if ( $this->isUserTalkPage() && $this->canSendUserTalkEmail() ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * @param array $watchers
+	 * @return bool
+	 */
+	private function thereAreUsersWatchingPage( array $watchers ) {
+		return count( $watchers ) || count( \F::app()->wg->UsersNotifiedOnAllChanges );
+	}
+
+	/**
+	 * @return bool
+	 */
+	private function canSendUserTalkEmail() {
+		$targetUser = User::newFromName( $this->title->getText() );
+		if ( !empty( \F::app()->wg->EnotifUserTalk ) && // we want to notify users when they're user talk page is changed
```

Comment:
Monster of a conditional!  Given this logic is the only thing this function does, I'd consider splitting these out to their own checks, and moving the explanation above them e.g.:
```php
// Should we notify users when their user talk page is changed?
if ( empty( \F::app()->wg->EnotifUserTalk ) ) {
  return false;
}
// The editor doesn't want users to be notified when they make minor edits on discussion pages
if ( $this->editor->isAllowed( 'nominornewtalk' ) ) {
  return false;
}
...
return true;
```
Reversing the logic like this also allows several negations to be removed, which can be easier to read.